### PR TITLE
Add initial anomaly_detection task with IsolationForest

### LINKED ISSUE #413: support anomaly detection
Add a task type 'anomaly_detection', and at least one learner for it. 

From my understanding there are multiple types of anomalies - point anomalies, contextual anomalies, and collective anomalies - and also different types of anomaly detection algorithms - nearest-neighbor, clustering, classification, and statistic based.

If anyone has any suggestions or requests for this task, please comment.

### PR DESCRIPTION
**Pull request by rashidrao-pk**
Closes #413

This draft PR adds initial support for anomaly detection in FLAML using IsolationForest.

Scope:
- Add `anomaly_detection` task type
- Add `is_anomaly_detection()`
- Register `isolation_forest`
- Add `IsolationForestEstimator` subclassing `SKLearnEstimator`
- Expose `predict`, `score_samples`, and `decision_function`
- Keep sklearn convention: `1` = normal, `-1` = anomaly
- Add unit test for synthetic point anomalies

Notes:
- This first PR intentionally keeps the scope minimal to lock the API.
- Full label-free `AutoML.fit(X_train=..., task="anomaly_detection")` integration can be completed after maintainer feedback.

cc @thinkall

### CODE REVIEW

**Comment by rashidrao-pk** [comment]
@microsoft-github-policy-service agree

**Comment by rashidrao-pk** [comment]
cc @int-chaos for review, as discussed in #413.

**Comment by rashidrao-pk** [comment]
Thanks for the review. I will fix the search-space init value, formatting, and Spark-dataframe error path.

For the `ap` metric comment: I agree this needs proper handling through anomaly scores (`score_samples`) in the AutoML evaluation path. Since this draft PR intentionally does not yet implement full label-free `AutoML.fit(...)` integration, I propose to address the metric evaluation path in the next PR together with the AutoML integration.

**Comment by thinkall** [comment]
Hi @rashidrao-pk , could you add an e2e test for anomaly_detection? Thanks a lot!

**Comment by rashidrao-pk** [comment]
Hi @thinkall,

Thanks for the suggestion.

I've added an end-to-end test for `anomaly_detection` and pushed the updates to this PR.

The test now exercises the `AutoML.fit(...)` workflow using the new `anomaly_detection` task and verifies that the estimator is correctly integrated into the AutoML pipeline, rather than only testing `IsolationForestEstimator` in isolation.

In addition, I reran the test suite and the project's pre-commit checks locally:

- ✅ `pytest test/automl/test_anomaly_detection.py`
- ✅ `pre-commit run --all-files`

Please let me know if you'd like the e2e test to cover any additional scenarios or edge cases.

**Comment by thinkall** [comment]
Hi @rashidrao-pk , could you help addressing the latest comments? Thanks.

**Comment by rashidrao-pk** [comment]
Hi @thinkall,

Thanks for the follow-up. I've addressed the latest review comments and pushed the updates.

The changes now include:

* anomaly-specific score handling for `ap` / `roc_auc` using continuous anomaly scores,
* holdout splitting support for `anomaly_detection`,
* preserving `n_jobs` for `IsolationForest`,
* removal of the unused estimator import,
* additional e2e coverage for `decision_function()`.

I also verified locally that:

* `pytest test/automl/test_anomaly_detection.py -vv` passes
* `pre-commit run --all-files` passes
* `git diff --check` is clean

Please let me know if any further adjustments are needed.

File: flaml/automl/task/generic_task.py [hunk 1]
```diff
@@ -1364,6 +1374,8 @@ def default_metric(self, metric: str) -> str:
             return "mape"
         elif self.is_rank():
             return "ndcg"
+        elif self.is_anomaly_detection():
+            return "ap"
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/automl/task/generic_task.py:1378]: Hi @rashidrao-pk ,  could you address this comment? Thanks.

File: flaml/automl/task/generic_task.py [hunk 2]
```diff
@@ -1089,6 +1091,10 @@ def decide_split_type(
             ), "GroupKFold requires groups to be provided."
             return split_type
 
+        elif self.is_anomaly_detection():
+            assert split_type in ["auto", "uniform", "time", "group"]
+            return split_type if split_type != "auto" else "uniform"
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/automl/task/generic_task.py:1096]: Hi @rashidrao-pk , could you address this comment? Thanks.

**Review by thinkall — CHANGES_REQUESTED** [review]
The anomaly-detection integration is not ready to merge yet:

1. The advertised label-free flow is unreachable: `AutoML.fit(X_train=X, task="anomaly_detection")` is rejected by generic validation before `IsolationForestEstimator.fit()` can ignore `y_train`. Define and implement X-only fitting semantics, including how model selection works without labeled validation data.
2. Explicit estimator lists are not task-validated, so classification accepts `isolation_forest` and anomaly detection accepts learners such as `rf`; both fail later through missing `predict_proba()`/`score_samples()` APIs.
3. Public `AutoML.score()` either calls nonexistent `IsolationForest.score()` or evaluates AP on hard `{-1, 1}` predictions, disagreeing with the continuous anomaly scores used during tuning.
4. `score_samples()` and `decision_function()` bypass the fitted `DataTransformer`; add AutoML-level forwarding APIs so they accept the same raw inputs as `predict()`.
5. Restrict or correctly implement explicit anomaly metrics. Several currently accepted metrics use inverted hard labels or unsupported probability APIs.
6. Strengthen the E2E test to verify anomaly ranking quality and cover X-only fitting/public scoring, not only output shape and label domain.

**Comment by rashidrao-pk** [comment]
Thanks for the detailed review. I’ve pushed an update addressing the requested anomaly-detection issues.

The changes now:

- support label-free AutoML.fit(X_train=X, task="anomaly_detection") using a one-shot initial configuration without fabricating labels;
- reject unlabeled HPO requests that would require a supervised validation objective;
- validate estimator/task compatibility in both directions;
- add public AutoML.score_samples() and decision_function() APIs with the same preprocessing path as predict();
- make AutoML.score() use continuous anomaly scores and preserve the fitted metric;
- restrict built-in anomaly metrics to ap and roc_auc;
- normalize both {0,1} and IsolationForest-style {-1,1} anomaly labels consistently;
- strengthen E2E tests for ranking quality, X-only fitting, public scoring, metric validation, and estimator compatibility.

The targeted regression suite currently passes:
25 passed, 1 deselected.

Commit: 5b892aad2

Thanks again for the review — happy to make any further adjustments.

**Review by thinkall — CHANGES_REQUESTED** [review]
I reviewed the commits since my last review. The earlier integration issues are substantially improved, but three blockers remain:

1. `flaml/automl/automl.py:2431`: single-class anomaly labels are still accepted for HPO, producing meaningless optimization. Validate the effective holdout/CV labels and require both normalized classes before search.
2. `flaml/automl/ml.py:653`: supported X-only training with labeled validation crashes when `log_training_metric=True` because training-label normalization receives `None`. Skip training-loss calculation when `y_train is None`.
3. `flaml/automl/ml.py:307`: string anomaly labels are generically label-encoded before anomaly normalization, which can invert their meaning and the resulting metric. Normalize/validate anomaly labels before generic encoding, or explicitly reject nonnumeric labels.

**Review by thinkall — CHANGES_REQUESTED** [review]
The merge from `main` is conflict-free but does not resolve the outstanding blockers:

1. `flaml/automl/automl.py:2516`: single-class anomaly labels still permit meaningless HPO instead of rejecting evaluation data without both classes.
2. `flaml/automl/ml.py:658`: X-only training with labeled validation and `log_training_metric=True` still attempts to normalize `y_train=None` and raises.
3. `flaml/automl/data.py:397`: nonnumeric anomaly labels are still generically label-encoded before anomaly normalization, which can reverse normal/anomaly semantics.

**Review by thinkall — CHANGES_REQUESTED** [review]
The merge from `main` is conflict-free but the outstanding blockers remain:

1. `flaml/automl/automl.py:2516-2520` and `flaml/automl/ml.py:303-308`: labeled anomaly HPO still accepts single-class labels, producing meaningless AP/ROC-AUC optimization.
2. `flaml/automl/ml.py:658-669`: X-only training with labeled validation and `log_training_metric=True` still normalizes `y_train=None` and raises.
3. `flaml/automl/data.py:397-406`: nonnumeric anomaly labels are still generically label-encoded before anomaly normalization, which can reverse normal/anomaly semantics.

**Review by thinkall — CHANGES_REQUESTED** [review]
Overall review of the complete current PR: changes are still required.

1. `flaml/automl/ml.py:303-308`: labeled anomaly HPO accepts single-class `{0}` or `{1}` evaluation labels, producing meaningless AP/ROC-AUC optimization. Require the normalized evaluation-label set to be exactly `{0, 1}`.
2. `flaml/automl/ml.py:658-669`: supported X-only training with labeled validation and `log_training_metric=True` still normalizes `y_train=None` and raises. Skip training-loss computation when training labels are absent.
3. `flaml/automl/data.py:397-406`: generic `LabelEncoder` processing can reverse nonnumeric anomaly-label semantics before normalization. Use an explicit anomaly mapping first or reject nonnumeric labels.

Posted by thinkall-agent-auto-reviewer

**Review by thinkall — CHANGES_REQUESTED** [review]
Overall review of the complete current PR: changes are still required.

1. `flaml/automl/automl.py:2516-2520` and `flaml/automl/ml.py:303-308`: single-class normalized anomaly labels still permit meaningless AP/ROC-AUC HPO. Require the effective evaluation-label set to be exactly `{0, 1}`.
2. `flaml/automl/ml.py:658-669`: X-only HPO with labeled validation and `log_training_metric=True` still calls anomaly-label normalization with `y_train=None`. Skip training-loss computation when labels are absent.
3. `flaml/automl/data.py:397-406` and `flaml/automl/automl.py:892-893`: generic label encoding reverses nonnumeric anomaly semantics and can make prediction fail on `-1`. Explicitly map anomaly labels before generic encoding, or reject nonnumeric labels.
4. `flaml/automl/automl.py:843-846` and `flaml/automl/model.py:1574-1577`: callable anomaly metrics can drive fitting but are ignored or rejected by public `AutoML.score()`. Either reject them during validation or support them consistently.

The current formatting workflow also fails because Black rewrites five changed files.

Posted by thinkall-agent-auto-reviewer

**Review by thinkall — CHANGES_REQUESTED** [review]
Overall review of the complete current PR: changes are still required.

1. `flaml/automl/ml.py:295-309`: single-class anomaly evaluation labels still permit meaningless AP/ROC-AUC HPO. Require the normalized evaluation-label set to be exactly `{0, 1}`.
2. `flaml/automl/ml.py:658-669`: X-only HPO with `log_training_metric=True` still normalizes `y_train=None` and raises. Skip training-loss calculation when training labels are absent.
3. `flaml/automl/data.py:397-406` and `flaml/automl/automl.py:892-893`: generic encoding reverses nonnumeric anomaly semantics and prediction can fail while inverse-transforming `-1`.
4. `flaml/automl/automl.py:843-846,2608-2613`: callable anomaly metrics can drive fitting but are not honored by public `AutoML.score()`.
5. Current formatting CI fails because Black rewrites five changed files.

Posted by thinkall-agent-auto-reviewer
